src/networkLayer/client.js

```diff
@@ -0,0 +1,206 @@
+/* @flow */
+import Relay from 'react-relay/classic';
+import RequestObject from './requestObject';
+
+export type ClientNetworkLayerOpts = {
+    graphqlUrl?: ?string,
+    headers?: ?Headers,
+    credentials?: ?string,
+    csrf?: ?boolean,
+    on403callback?: ?() => void,
+    onSpeedbarResponse?: ?(response: any) => void,
+    onJailedCallback?: ?() => void,
+    onMutationStartCallback?: ?(id: number, data?: Object) => void,
+    onMutationProgressCallback?: ?(id: number, received: number, size: number) => void,
+    env: Relay.Environment
+};
+
+var nextRequestId = 0;
+var nextTraceId = 0;
+
+function addTraceIds(trace) {
+    const keys = Object.keys(trace).filter(key => key !== '__trace');
+    if (trace.__trace) {
+        trace.__clientTraceId = nextTraceId++;
+    }
+    if (keys.length) {
+        keys.forEach(key => {
+            addTraceIds(trace[key]);
+        });
+    }
+    return trace;
+}
+
+const getUniqueRequestID = (request) => request.getID
+    ? request.getID()
+    : request.getVariables().input_0.clientMutationId;
+
+export default class ClientNetworkLayer {
+    url = '/graphql';
+    headers = {
+        'X-Requested-With': 'XMLHttpRequest',
+        'Content-Type': 'application/json',
+    };
+    credentials = 'same-origin';
+    csrf = true;
+    on403callback = null;
+    onSpeedbarResponse = null;
+    onJailedCallback = null;
+    env: Relay.Environment;
+
+    supports() {}
+
+    sendQueries(queryRequests: any) {
+        return Promise.all(queryRequests.map(this.sendQuery));
+    }
+
+    sendQuery = (queryRequest: any) => {
+        return this._send(queryRequest.getID(), 'query', queryRequest);
+    };
+
+    sendMutation(mutationRequest: any) {
+        const id = getUniqueRequestID(mutationRequest);
+        return this._send(id, 'mutation', mutationRequest);
+    }
+
+    // We store the XHR as a RequestObject so that it can be retrieved and manipulated later
+    _storeRequest = (request: XMLHttpRequest, id: string) => {
+        const removeSelf = () => {
+            this.env._requests.delete(id);
+        };
+
+        const endOfLifeEvents = ['error', 'abort', 'loadend', 'timeout'];
```

Comment:
I thought `load` fired when all data had been successfully read, whereas `loadend` fires when the data has finished transferring, and it always fires after error abort or load, so it's like a safety catch-all.

https://developer.mozilla.org/en-US/docs/Web/Events/loadend
https://www.nczonline.net/blog/2012/05/22/working-with-files-in-javascript-part-3/